attestation/jsonattestor/json.go

```diff
@@ -0,0 +1,213 @@
+package jsonattestor
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/cyberphone/json-canonicalization/go/src/webpki.org/jsoncanonicalizer"
+	"github.com/ghodss/yaml"
+	"github.com/itchyny/gojq"
+
+	"github.com/in-toto/go-witness/attestation"
+	"github.com/in-toto/go-witness/cryptoutil"
+	"github.com/in-toto/go-witness/log"
+	"github.com/in-toto/go-witness/registry"
+	"github.com/invopop/jsonschema"
+)
+
+const (
+	Name    = "json"
+	Type    = "https://witness.dev/attestations/json/v0.1"
+	RunType = attestation.PostProductRunType
+)
+
+// Attestor processes a JSON/YAML file, storing canonical JSON and user queries.
+type Attestor struct {
+	CanonicalJSON  []byte
+	SubjectQueries map[string]string
+	subjects       map[string]cryptoutil.DigestSet
+}
+
+var (
+	_ attestation.Attestor  = &Attestor{}
+	_ attestation.Subjecter = &Attestor{}
+)
+
+func init() {
+	attestation.RegisterAttestation(
+		Name,
+		Type,
+		RunType,
+		func() attestation.Attestor { return New() },
+		registry.StringConfigOption(
+			"subject-queries",
+			"subjectName=jq expression pairs (semicolon-separated). Example: 'foo=.foo;bar=.nums[]'",
+			"",
+			func(a attestation.Attestor, sqStr string) (attestation.Attestor, error) {
+				att, ok := a.(*Attestor)
+				if !ok {
+					return a, fmt.Errorf("unexpected attestor type: %T", a)
+				}
+				sqMap, err := ParseSubjectQueries(sqStr)
+				if err != nil {
+					return a, fmt.Errorf("parseSubjectQueries: %w", err)
+				}
+				att.SubjectQueries = sqMap
+				return att, nil
+			},
+		),
+	)
+}
+
+func New() *Attestor {
+	return &Attestor{
+		subjects:       make(map[string]cryptoutil.DigestSet),
+		SubjectQueries: make(map[string]string),
+	}
+}
+
+func ParseSubjectQueries(sqStr string) (map[string]string, error) {
+	out := make(map[string]string)
+	s := strings.TrimSpace(sqStr)
+	if s == "" {
+		return out, nil
+	}
+	pairs := strings.Split(s, ";")
+	for _, pair := range pairs {
+		pair = strings.TrimSpace(pair)
+		if pair == "" {
+			continue
+		}
+		parts := strings.SplitN(pair, "=", 2)
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("invalid subject-queries pair %q (missing '=')", pair)
+		}
+		name := strings.TrimSpace(parts[0])
+		expr := strings.TrimSpace(parts[1])
+		if name == "" || expr == "" {
+			return nil, fmt.Errorf("invalid subject-queries pair %q (empty name or expression)", pair)
+		}
+		out[name] = expr
+	}
+	return out, nil
+}
+
+func (a *Attestor) Name() string {
+	return Name
+}
+
+func (a *Attestor) Type() string {
+	return Type
+}
+
+func (a *Attestor) RunType() attestation.RunType {
+	return RunType
+}
+
+func (a *Attestor) Schema() *jsonschema.Schema {
+	return jsonschema.Reflect(a)
+}
+
+func (a *Attestor) Attest(ctx *attestation.AttestationContext) error {
+	prods := ctx.Products()
+	if len(prods) == 0 {
+		return errors.New("no products to attest")
+	}
+	path := pickOneJSONFile(prods)
```

Comment:
might become clear while reviewing further, but why are we only attesting one json file 👀

Reply:
I wanted to avoid multi-doc and multi-file issues and edge cases there are alot.  I have solved all that I can think of now